### PR TITLE
React to externally left and joined rooms, and add "leave room" button in room menu

### LINKED ISSUE #71: No Feedback When Joining New Channels
Related to #70. When typing commands, the only way I knew I was actually joining channels was through the chime the riot app makes on my phone. Perhaps a similar chime could be worked out?

### PR DESCRIPTION
**Pull request by vurpo**
Fixes #71, #72, and partially #25.

There is now a "Leave room" option in the room menu (top-right corner, and in the right-click menu in the roomlist), which leaves the room.

Also, nheko now adds/removes rooms from the room list and room view when you join and leave rooms from another client.

### CODE REVIEW

**Comment by mujx** [comment]
The build fails on `make lint`. You just need to run it locally to apply the style changes.

**Comment by vurpo** [comment]
I'll fix this tomorrow, thanks

**Comment by vurpo** [comment]
I don't get how `make lint` is supposed to work, I ran it and it seemed to change every file, and the CI build still failed on `make lint`

**Comment by vurpo** [comment]
I removed my `make lint` commit for now, because it seemed to go very wrong somehow.

**Comment by vurpo** [comment]
hmmm, I got clang-format v5.0.0 and it seemed to correctly detect the clang-format configuration, but it still changed something that didn't go through the lint check in CI

File: src/RoomList.cc [hunk 1]
```diff
@@ -72,6 +92,34 @@ RoomList::clear()
...
+RoomList::addRoom(const QSharedPointer<RoomSettings> &settings,
+                  const RoomState &state,
+                  const QString &room_id)
+{
+        RoomInfoListItem *room_item = new RoomInfoListItem(settings, state, room_id, scrollArea_);
+        connect(room_item, &RoomInfoListItem::clicked, this, &RoomList::highlightSelectedRoom);
+        connect(room_item, &RoomInfoListItem::leaveRoom, client_.data(), &MatrixClient::leaveRoom);
+
+        rooms_.insert(room_id, QSharedPointer<RoomInfoListItem>(room_item));
+
```
Comment (mujx) [inline src/RoomList.cc]: You will need to fetch the room avatar with `MatrixClient::fetchRoomAvatar`

File: src/RoomList.cc [hunk 2]
```diff
@@ -118,6 +160,19 @@ RoomList::setInitialRooms(const QMap<QString, QSharedPointer<RoomSettings>> &set
                   new RoomInfoListItem(settings[room_id], state, room_id, scrollArea_);
                 connect(
                   room_item, &RoomInfoListItem::clicked, this, &RoomList::highlightSelectedRoom);
+                connect(room_item, &RoomInfoListItem::leaveRoom, client_.data(), [=]() {
```
Comment (mujx) [inline src/RoomList.cc]: You can move this lambda into a function so it's not duplicated.

File: src/RoomList.cc [hunk 3]
```diff
@@ -118,6 +160,19 @@ RoomList::setInitialRooms(const QMap<QString, QSharedPointer<RoomSettings>> &set
...
+                        leaveRoomDialog_ = new LeaveRoomDialog(this);
+                        connect(leaveRoomDialog_,
+                                &LeaveRoomDialog::closing,
+                                this,
+                                [=](bool leaving) { closeLeaveRoomDialog(leaving, room_id); });
+
+                        leaveRoomModal = new OverlayModal(MainWindow::instance(), leaveRoomDialog_);
+                        leaveRoomModal->setDuration(100);
```
Comment (mujx) [inline src/RoomList.cc]: Also set the duration to 0.

**Comment by rokups** [comment]
When i try to leave room i get spam of this:
```
     5.116: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 11284, resource id: 83886291, major code: 130 (Unknown), minor code: 3
     5.116: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 11295, resource id: 83886291, major code: 130 (Unknown), minor code: 3
     5.120: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 11426, resource id: 83886291, major code: 130 (Unknown), minor code: 3
     5.146: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 11592, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    17.106: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 12810, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    18.525: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 12986, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    22.046: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13178, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    23.590: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13354, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.805: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13429, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.820: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13470, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.823: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13511, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.839: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13552, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.844: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13598, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.855: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13639, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.871: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13680, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.887: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13721, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.902: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13762, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.919: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13803, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.935: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13844, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.951: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13885, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.967: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13926, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.983: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 13967, resource id: 83886291, major code: 130 (Unknown), minor code: 3
    27.999: [warning] - QXcbConnection: XCB error: 8 (BadMatch), sequence: 14008, resource id: 83886291, major code: 130 (Unknown), minor code: 3
```

UI freezes with a undarkened rect in the middle (confirmation window?) while rest of window is darkened. Have to restart. After restart room is still there.

**Comment by vurpo** [comment]
That's weird, does it also happen if you try to log out of nheko?

**Comment by mujx** [comment]
@rokups Better open a new issue for that. I think it's related to the overlay because I have seen it happen before on KDE.

**Pull request merged**